### ISSUE TITLE
Enable leak tracking for release mode.

### ISSUE THREAD

**Issue by polina-c** [dart-lang/leak_tracker#25]
Subtasks:

- [x] Make `reachabilityBarrier` (package `dart:developer`) available in release mode: https://github.com/itsjustkevin/sdk/commit/08b76a50fc138a94a5460f321500911b666eada1
- [ ] Enable leak tracker for release mode.

If you are interested in leak tracking in release mode, please, upvote this issue and justify  your case.

**Comment by marandaneto** [comment]
I believe this would be very helpful for release builds as well, so people can upload their leaks to error reporting tools such as sentry.io and fix them.
Similar to https://square.github.io/leakcanary/uploading/

**Comment by noga-dev** [comment]
Could be useful for edge cases with special releases on beta channel. Would be good to add an ```enableInRelease``` flag.

**Comment by polina-c** [comment]
@a-siva and @bkonyi , thoughts?

**Comment by a-siva** [comment]
> @a-siva and @bkonyi , thoughts

This is for flutter release builds I presume or is this request for command line Dart ?

**Comment by bkonyi** [comment]
I assume this is for `PRODUCT` mode builds, which includes Flutter release. Given that this is basically a counter that we increment (even in `PRODUCT` mode, although it's not accessible) I think it's low-risk to expose this in `PRODUCT` mode. However, if this feature requires any sort of VM service connection to work properly then this won't be something we can do.

**Comment by polina-c** [comment]
Thanks. Yes, it is just counter. VM service connection is not needed as the diagnostics will be saved and analyzed offline. Can we plan implementing this? Should I transfer the issue to VM repo?

**Comment by fzyzcjy** [comment]
I vote for it, if the performance drop is minimal in release mode. More sanity checks never do any harm (as long as they do not affect performance)!

**Comment by a-siva** [comment]
https://dart-review.googlesource.com/c/sdk/+/315981

**Comment by ZhouC125** [comment]
> I vote for it, if the performance drop is minimal in release mode. More sanity checks never do any harm (as long as they do not affect performance)!

I agree with this comment.

**Comment by polina-c** [comment]
The feature is unblocked, but not implemented yet, and not part of leak_tracker MVP.
Reopening.

**Comment by a-siva** [comment]
> The feature is unblocked, but not implemented yet, and not part of leak_tracker MVP. Reopening.

We have turned  the reachability barrier on in product mode, why do you say it is not implemented yet?

**Comment by polina-c** [comment]
@itsjustkevin, I am reopening the issue: while reachability barrier is now available in release mode, leak tracker is still off for release mode.

**Comment by dnfield** [comment]
I do not think we should make this an option unless we have significant profiling data to support it.

In particular, we should have profiling data from a mid-tier Android device that shows this has little to no impact on frame/raster times. The profiling data we do have so far from macOS desktop applications suggests that it would.

**Comment by polina-c** [comment]
> I do not think we should make this an option unless we have significant profiling data to support it.
> 
> In particular, we should have profiling data from a mid-tier Android device that shows this has little to no impact on frame/raster times. The profiling data we do have so far from macOS desktop applications suggests that it would.

I agree. Most likely in release mode leak analysis and reporting should happen not on timer, but on request from application, when application is known to be in stale state.

**Comment by polina-c** [comment]
The Dash team does not have bandwidth for this in the nearest future, but

**Comment by polina-c** [comment]
In fact, leak tracking can be easily enabled for profile mode by setting flag `--dart-define=flutter.memory_allocations=true`.

So, to get leak tracking enabled in your prod applications, you need to:
1. Compile it in profile mode with the flag above
2. Configure leak tracker not to report leaks to console, but instead report them to some storage, local or external, making sure (1) it is aligned with your privacy and security policies and (2) it does not impact the app performance.
